src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

```diff
@@ -548,55 +540,37 @@ Task HandlePreAuthentication (HttpURLConnection httpConnection)
 			if (Logger.LogNet)
 				Logger.Log (LogLevel.Info, LOG_APP, $"Authentication header '{data.UseProxyAuthentication ? "Proxy-Authorization" : "Authorization"}' will be set to '{authorization.Message}'");
 			httpConnection.SetRequestProperty (data.UseProxyAuthentication ? "Proxy-Authorization" : "Authorization", authorization.Message);
-			});
 		}
 
-		Task AddHeaders (HttpURLConnection conn, HttpHeaders headers)
+		void AddHeaders (HttpURLConnection conn, HttpHeaders headers)
 		{
-			return Task.Factory.StartNew (() => {
 			if (headers == null)
 				return;
 
 			foreach (KeyValuePair<string, IEnumerable<string>> header in headers) {
 				conn.SetRequestProperty (header.Key, header.Value != null ? String.Join (",", header.Value) : String.Empty);
 			}
-			});
 		}
 		
-		async Task SetupRequestBody (HttpURLConnection httpConnection, HttpRequestMessage request)
+		void SetupRequestBody (HttpURLConnection conn, HttpRequestMessage request)
 		{
 			if (request.Content == null) {
-				httpConnection.SetChunkedStreamingMode (0);
 				// Pilfered from System.Net.Http.HttpClientHandler:SendAync
 				if (HttpMethod.Post.Equals (request.Method) || HttpMethod.Put.Equals (request.Method) || HttpMethod.Delete.Equals (request.Method)) {
 					// Explicitly set this to make sure we're sending a "Content-Length: 0" header.
 					// This fixes the issue that's been reported on the forums:
 					// http://forums.xamarin.com/discussion/17770/length-required-error-in-http-post-since-latest-release
-					httpConnection.SetRequestProperty ("Content-Length", "0");
+					conn.SetRequestProperty ("Content-Length", "0");
 				}
 				return;
 			}
 			
-			httpConnection.DoOutput = true;
-
-			var bytes = await request.Content.ReadAsByteArrayAsync ().ConfigureAwait (false);
-
-			int contentLength = bytes.Length;
-			httpConnection.SetRequestProperty ("Content-Length", contentLength.ToString());
-			httpConnection.SetFixedLengthStreamingMode (contentLength);
-
-			string contentType;
-			if (request.Content.Headers.ContentType != null)
-				contentType = String.Join (" ", request.Content.Headers.GetValues ("Content-Type"));
+			conn.DoOutput = true;
+			long? contentLength = request.Content.Headers.ContentLength;
+			if (contentLength != null)
+				conn.SetFixedLengthStreamingMode ((int)contentLength);
 			else
-				contentType = "text/plain";
-			httpConnection.SetRequestProperty ("Content-Type", contentType);
-
-			await Task.Factory.StartNew (() => {
-				httpConnection.OutputStream.Write (bytes, 0, contentLength);
-				httpConnection.OutputStream.Flush ();
-				httpConnection.OutputStream.Close ();
-			});
+				conn.SetChunkedStreamingMode (0);
```

Comment:
Why are you removing the `Content-Length` header from the request? This code was a fix to POST problems IIRC. The same appliesto `Content-Type`. And the same applies to writing and flushing the content - without this code POST would sometimes fail to work. @blounty, can you chime in please? You fixed the POST bug